kernel/ixgbe/src/lib.rs

```diff
@@ -0,0 +1,1192 @@
+                    BoxRefMut<MappedPages, IntelIxgbeRegisters3>, 
+                    BoxRefMut<MappedPages, IntelIxgbeMacRegisters>, 
+                    Vec<IxgbeRxQueueRegisters>, Vec<IxgbeTxQueueRegisters>), &'static str> 
+    {
+        // We've divided the memory-mapped registers into multiple regions.
+        // The size of each region is found from the data sheet, but it always lies on a page boundary.
+        const GENERAL_REGISTERS_1_SIZE_BYTES:   usize = 4096;
+        const RX_REGISTERS_SIZE_BYTES:          usize = 4096;
+        const GENERAL_REGISTERS_2_SIZE_BYTES:   usize = 4 * 4096;
+        const TX_REGISTERS_SIZE_BYTES:          usize = 2 * 4096;
+        const MAC_REGISTERS_SIZE_BYTES:         usize = 5 * 4096;
+        const GENERAL_REGISTERS_3_SIZE_BYTES:   usize = 18 * 4096;
+
+        // Allocate memory for the registers, making sure each successive memory region begins where the previous region ended.
+        let nic_regs1_mapped_page = allocate_memory(mem_base, GENERAL_REGISTERS_1_SIZE_BYTES)?;
+        let nic_rx_regs1_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES, RX_REGISTERS_SIZE_BYTES)?;
+        let nic_regs2_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES, GENERAL_REGISTERS_2_SIZE_BYTES)?;        
+        let nic_tx_regs_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES, TX_REGISTERS_SIZE_BYTES)?;
+        let nic_mac_regs_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES, MAC_REGISTERS_SIZE_BYTES)?;
+        let nic_rx_regs2_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES + MAC_REGISTERS_SIZE_BYTES, RX_REGISTERS_SIZE_BYTES)?;        
+        let nic_regs3_mapped_page = allocate_memory(mem_base + GENERAL_REGISTERS_1_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES + GENERAL_REGISTERS_2_SIZE_BYTES + TX_REGISTERS_SIZE_BYTES + MAC_REGISTERS_SIZE_BYTES + RX_REGISTERS_SIZE_BYTES, GENERAL_REGISTERS_3_SIZE_BYTES)?;
+
+        // Map the memory as the register struct and tie the lifetime of the struct with its backing mapped pages
+        let regs1 = BoxRefMut::new(Box::new(nic_regs1_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters1>(0))?;
+        let regs2 = BoxRefMut::new(Box::new(nic_regs2_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters2>(0))?;
+        let regs3 = BoxRefMut::new(Box::new(nic_regs3_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeRegisters3>(0))?;
+        let mac_regs = BoxRefMut::new(Box::new(nic_mac_regs_mapped_page)).try_map_mut(|mp| mp.as_type_mut::<IntelIxgbeMacRegisters>(0))?;
+        
+        // Divide the pages of the Rx queue registers into multiple 64B regions
```

Comment:
Since the Rx and Tx register sets are 64B each, it's almost like we have to create a special heap for the NIC that can track these allocations from the backing MappedPages.